File: Monotonic/StockSpanner.js
class StockSpanner {
  constructor() {
    // Monotonic decreasing stack to store [price, span] pairs
    this.priceStack = [];
  }

  next(price) {
    // Stock span is always at least 1
    let stockSpanDays = 1;
    while (this.priceStack.length && this.priceStack.at(-1)[0] <= price) {
      let prevSpan = this.priceStack.pop()[1];
      stockSpanDays += prevSpan;
    }

    this.priceStack.push([price, stockSpanDays]);

    return stockSpanDays;
  }
}

// Constant time complexity for each next call, linear space

const stockSpanner = new StockSpanner();

console.log(stockSpanner.next(100));
console.log(stockSpanner.next(80));
console.log(stockSpanner.next(60));
console.log(stockSpanner.next(70));
console.log(stockSpanner.next(60));
console.log(stockSpanner.next(75));
console.log(stockSpanner.next(85));
console.log(stockSpanner.priceStack);

File: Monotonic/StockSpanner.java

import java.util.ArrayDeque;
import java.util.Deque;

public class StockSpanner {
    Deque<int[]> priceStack;

    public StockSpanner() {
        // Monotonic decreasing stack to store price and span days
        this.priceStack = new ArrayDeque<>();
    }

    public int next(int price) {
        int span = 1;
        while (!priceStack.isEmpty() && priceStack.getLast()[0] <= price) {
            span += priceStack.removeLast()[1];
        }

        priceStack.addLast(new int[] { price, span });

        return span;
    }

    public static void main(String[] args) {
        StockSpanner obj = new StockSpanner();
        System.out.println(obj.next(100));
        System.out.println(obj.next(80));
        System.out.println(obj.next(60));
        System.out.println(obj.next(70));
        System.out.println(obj.next(60));
        System.out.println(obj.next(75));
        System.out.println(obj.next(85));
        System.out.println(obj.priceStack);
    }
}

// Constant time complexity for each next call, linear space complexity for the stack in the worst case (strictly decreasing prices)